pygfx/objects/_events.py

```diff
@@ -0,0 +1,310 @@
+            return _callback
+
+        if decorating:
+            return decorator
+        return decorator(callback)
+
+    def remove_event_handler(self, callback, *types):
+        """Unregister an event handler.
+
+        Arguments:
+            callback (callable): The event handler.
+            *types (list of strings): A list of event types.
+        """
+        for type in types:
+            self._event_handlers[type].remove(callback)
+
+    def handle_event(self, event: Event):
+        """Handle an incoming event.
+
+        Subclasses can overload this method. Events include widget
+        resize, mouse/touch interaction, key events, and more. An event
+        is a dict with at least the key event_type. For details, see
+        https://jupyter-rfb.readthedocs.io/en/latest/events.html
+        """
+        event_type = event.type
+        for callback in self._event_handlers[event_type]:
+            if event.is_cancelled:
+                break
+            callback(event)
+
+    def set_pointer_capture(self, pointer_id):
+        """Register this object to capture any other pointer events,
+        until ``release_pointer_capture`` is called or an ``pointer_up``
+        event is encountered.
+        """
+        EventTarget.pointer_captures[pointer_id] = self
+
+    def release_pointer_capture(self, pointer_id):
+        """Release the pointer capture for the object that was registered
+        to the given pointer_id.
+        """
+        if pointer_id in EventTarget.pointer_captures:
+            del EventTarget.pointer_captures[pointer_id]
+
+
+class RootEventHandler(EventTarget):
+    """Root event handler for the Pygfx event system."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def dispatch_event(self, event: dict):
+        # Check for captured pointer events
+        pointer_id = getattr(event, "pointer_id", None)
+        if pointer_id is not None and pointer_id in EventTarget.pointer_captures:
+            captured_target = EventTarget.pointer_captures[pointer_id]
+            # Set the target to be the captured target
+            event._target = captured_target
+            captured_target.handle_event(event)
+            if event.type == EventType.POINTER_UP:
+                captured_target.release_pointer_capture(pointer_id)
+            # Encountered an event with pointer_id while there is a
+            # capture active, so don't bubble, just return immediately
+            return
+
+        target = event.target
+        while target and target is not self:
+            # Update the private current target field
+            event._current_target = target
+            target.handle_event(event)
+            if pointer_id is not None and pointer_id in EventTarget.pointer_captures:
+                # Prevent people from shooting in their foot by calling set_pointer_capture
+                # on POINTER_UP events
```

Comment:
I'm having second thoughts about this here, but I do think it's good that we prevent any and all dangling pointer captures